crates/biome_formatter/src/printer/mod.rs

```diff
@@ -841,9 +872,12 @@ impl Indention {
             },
 
             // Convert the existing align to an indent
-            Indention::Align { level: indent, .. } => Indention::Align {
-                level: indent + 1,
-                align: count,
+            Indention::Align {
+                level: indent,
+                align,
+            } => Indention::Align {
+                level: indent,
+                align: align.saturating_add(count.get()),
```

Comment:
This issue happen when indent_style is `space` and indent_width is not 2
For example 
If you run the code below:

> This code is a part of https://github.com/biomejs/biome/pull/2134/commits/f7e6d79f53d81bc5716f93ddb9f406c0e396e467# code.

```
    let context = SimpleFormatContext::new(SimpleFormatOptions {
        indent_width: IndentWidth::try_from(8).unwrap(),
        indent_style: IndentStyle::Space,
        ..SimpleFormatOptions::default()
    });

    let elements = format!(
        context,
        [
            text("root"),
            align(
                2,
                &format_args![
                    hard_line_break(),
                    text("Aligned"),
                    indent(&format_args![
                        hard_line_break(),
                        text("Aligned, and indented"),
                        dedent(&format_args![
                            hard_line_break(),
                            text("aligned, not Intended"),
                        ]),
                    ])
                ]
            ),
            dedent(&format_args![hard_line_break(), text("root level")])
        ]
    )?;
```
`Previously  biome` result is
```
root
  Aligned
          Aligned, and indented
        aligned, not Intended
root level
```
`this branch` result is
```
root
  Aligned
          Aligned, and indented
  aligned, not Intended
root level
```

And I may not understand the entire prettier logic, but I think prettier also record align and indent state
https://github.com/prettier/prettier/blob/d970deed6b6cabe21fdb35249265d766932b5718/src/document/printer.js#L67-L70

Feel free to share your thoughts with me anytime.

Reply:
This PR truly fixes the problem you mentioned, but it might alter the behavior when `indent_style` is `tab`. I had some observations when solving https://github.com/biomejs/biome/pull/1581 

### Biome behavior
- Biome only keeps the most inner `Align`, when encountering another `Align`, Biome will automatically convert the preceding `Align` to `Indent`.
https://github.com/biomejs/biome/blob/62fbec86467946e8d71dcd0098e3d90443208f78/crates/biome_formatter/src/printer/mod.rs#L844-L847
  Let us use [playground](https://biomejs.dev/playground/?indentStyle=space&indentWidth=0&code=YQAKACAAIAAgACAAPwAgAC8ALwAgAHMAbwBtAGUAdABoAGkAbgBnACAAbwB1AHQAZQByAAoAIAAgACAAIAAgACAAeAAKACAAIAAgACAAOgAgAC8ALwAgAGUAbABzAGUAIABvAHUAdABlAHIACgAgACAAIAAgACAAIABiAAoAIAAgACAAIAAgACAAPwAgAC8ALwAgAHMAbwBtAGUAdABoAGkAbgBnACAAaQBuAG4AZQByAAoAIAAgACAAIAAgACAAIAAgAHkACgAgACAAIAAgACAAIAA6ACAALwAvACAAZQBsAHMAZQAgAGkAbgBuAGUAcgAKACAAIAAgACAAIAAgACAAIAB6ADsACgA%3D)(I adjust indent width to 0 for easy understanding) as an example
  ```
  // indent_style: space
  // for b
  [Indent, Align] => [Indent, Align]
  // for y 
  [Indent, Align, Align] => [Indent, Indent, Align]
  ```
  I guess the reason the fix works is you prevent `Align` from automatically converting to `Indent`. There is no `dedent` in the formatter IR.

- When `indent_style` is `tab`, the most inner `Align`(if exists) will be automatically converted to `Indent` when encountering `Indent`
  https://github.com/biomejs/biome/blob/c35697124a95fc42a11481ecfa7bc534ae8c9e50/crates/biome_formatter/src/printer/mod.rs#L810

  ```
  // indent_style: tab
  [indent, Align, Indent] => [Indent, Indent, Indent]
  ```

### Prettier behavior
- Prettier keeps all `Align`s when `indent_style` is space
- Prettier converts all preceding `Align`s to `Indents` when encountering `Indent` if `indent_style` is `tab`
  ```
  [Indent, Align, Align] => [Indent, Align, Align]
  [Indent, Align, Align, Indent] => [indent, Indent, Indent, Indent]
  [Indent, Align, Align, Indent, Dedent] => [Indent, Align, Align]
  [Indent, Align, Align, Indent, Indent, Dedent] => [Indent, Indent, Indent, Indent]
  ```
  You can see Prettier record the number of `Align`s
  https://github.com/prettier/prettier/blob/d970deed6b6cabe21fdb35249265d766932b5718/src/document/printer.js#L93
  And convert these `Align`s to `Indent`s when encountering `Indent`
  https://github.com/prettier/prettier/blob/d970deed6b6cabe21fdb35249265d766932b5718/src/document/printer.js#L80
 

This is all I can remember, and I'm not sure if it's correct or not, but I hope it helps you.